### PR TITLE
Improve squash merge commit author and co-author with private emails

### PR DESCRIPTION
**Pull request by brechtvl**
When emails addresses are private, squash merges always use `@noreply.localhost` for the author of the squash commit. And the author is redundantly added as a co-author in the commit message.

Also without private mails, the redundant co-author is possible when committing with a signature that's different than the user full name and primary email.

Now try to find a commit by the same user in the list of commits, and prefer the signature from that over one constructed from the account settings.

### CODE REVIEW

**Review by techknowlogick — CHANGES_REQUESTED** [review]
Thanks for the PR :)
LGTM, just need to confirm it doesn't potentially leak private emails from web interface prior to approval.

**Comment by brechtvl** [comment]
Thanks. And to state the obvious, the signature used in commits is already public information and preserved when using a merge commit or rebase as the merge style. So using it for the squash case as well does not expose private information.

**Review by techknowlogick — APPROVED** [review]

**Comment by techknowlogick** [comment]
Thanks for your patience @brechtvl 🙏

File: services/pull/merge_squash.go [hunk 1]
```diff
@@ -7,24 +7,58 @@ import (
...
+	defer closer.Close()
+
+	commits, err := gitRepo.CommitsBetweenIDs(trackingBranch, "HEAD")
+	if err != nil {
+		return nil, fmt.Errorf("Unable to get commits between: %s %s Error %v", "HEAD", trackingBranch, err)
+	}
+
+	for _, commit := range commits {
+		if commit.Author != nil {
+			commitUser, _ := user_model.GetUserByEmail(ctx, commit.Author.Email)
```
Comment (lunny) [inline services/pull/merge_squash.go:43]: We could cache the users by email so that we could not load it from database again.

**Review by lunny — APPROVED** [review]

**Comment by wxiaoguang** [comment]
@brechtvl would you like to take a look at   Squash in PR with co-authored #31980  and Fix duplicate co-author in squashed merge commit messages #33020  ?

**Pull request merged**